### PR TITLE
rabbitmq delayed message exchange

### PR DESCRIPTION
**Pull request by barkhashah**
This change implements rabbitmq delayed message exchange. You will be required to enable the plugin rabbitmq_delayed_message_exchange. Added a member in the constructor that decides if the user wants to create and actual exchange or the delayed exchange.

### CODE REVIEW

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 1]
```diff
@@ -117,21 +131,22 @@ public class RabbitMQMessageBus : MessageBusBase, IMessageBus {
...
                 Data = await _serializer.SerializeToStringAsync(message).AnyContext()
             }).AnyContext();
             
             var basicProperties = _publisherChannel.CreateBasicProperties();
             basicProperties.Persistent = _persistent;
-            basicProperties.Expiration = _defaultMessageTimeToLive.Milliseconds.ToString();
-
+            basicProperties.Expiration = _defaultMessageTimeToLive.TotalMilliseconds.ToString(CultureInfo.InvariantCulture);
+            // RabbitMQ only supports delayed messages with a third party plugin called "rabbitmq_delayed_message_exchange"
+            if (_delayedExchange && delay.HasValue && delay.Value > TimeSpan.Zero) {
```
Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Is there any way to tell if this plugin is installed a head of time and throw an exception?? Or what will happen here if the plugin is not installed?

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 2]
```diff
@@ -117,21 +131,22 @@ public class RabbitMQMessageBus : MessageBusBase, IMessageBus {
                 return;
             _logger.Trace("Message Publish: {messageType}", messageType.FullName);
 
-            // RabbitMQ only supports delayed messages with a third party plugin.
-            if (delay.HasValue && delay.Value > TimeSpan.Zero) {
```
Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: I think this should still happen if you don't have _delayExchange set to true.. This will at least give you some kind of basic delayed message (as long as your app doesn't crash).

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 3]
```diff
@@ -105,7 +119,7 @@ public class RabbitMQMessageBus : MessageBusBase, IMessageBus {
         /// </summary>
         /// <param name="messageType"></param>
         /// <param name="message"></param>
-        /// <param name="delay"></param>
+        /// <param name="delay">Along with the delay value, _delayExchange should also be set to true</param>
```
Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: This shouldn't be required as we do have a base class to do delayed messages (in process) while not as good or reliable, this is an implementation detail :)

**Comment by niemyjski** [comment]
Thanks for the commit, I reviewed it and left a few comments. I think we should fall back to using the base class for delayed messaging.

Also, if you want to merge into master instead of the .net branch your changes will go live sooner. There was a time period where dotnet branch was master and we changed it around and merged in all the non .net core changes into master.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs (hunk 1 above)
Comment (barkhashah) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Well I tried to look for an api to find out if there is a way to know if plugin is installed but couldn't find one. But yes, if the plugin is not installed and you are trying to create x-delayed-message exchange then the exception gets thrown.

Unhandled Exception: RabbitMQ.Client.Exceptions.OperationInterruptedException: T
he AMQP operation was interrupted: AMQP close-reason, initiated by Peer, code=50
3, text="COMMAND_INVALID - invalid exchange type 'x-delayed-message'", classId=4
0, methodId=10, cause=
   at RabbitMQ.Client.Impl.SimpleBlockingRpcContinuation.GetReply(TimeSpan timeo
ut)
   at RabbitMQ.Client.Impl.ModelBase.ModelRpc(MethodBase method, ContentHeaderBa
se header, Byte[] body)
   at RabbitMQ.Client.Framing.Impl.Model._Private_ExchangeDeclare(String exchang
e, String type, Boolean passive, Boolean durable, Boolean autoDelete, Boolean in
ternal, Boolean nowait, IDictionary`2 arguments)
   at RabbitMQ.Client.Impl.ModelBase.ExchangeDeclare(String exchange, String typ
e, Boolean durable, Boolean autoDelete, IDictionary`2 arguments)
   at Foundatio.Messaging.RabbitMQMessageBus.SetUpExchangeAndQueuesForRouting(IM
odel model) in C:\Users\barkha.shah\Foundatio\src\Foundatio.RabbitMQ\Messaging\R
abbitMQMessageBus.cs:line 98
   at Foundatio.Messaging.RabbitMQMessageBus..ctor(String userName, String passw
ord, String queueName, String routingKey, String exhangeName, Boolean delayedExc
hange, Boolean durable, Boolean persistent, Boolean exclusive, Boolean autoDelet
e, IDictionary`2 queueArguments, Nullable`1 defaultMessageTimeToLive, ISerialize
r serializer, ILoggerFactory loggerFactory) in C:\Users\barkha.shah\Foundatio\sr
c\Foundatio.RabbitMQ\Messaging\RabbitMQMessageBus.cs:line 72
   at Foundatio.RabbitMQSubscribeConsole.Program.Main(String[] args)
Press any key to continue . . .

Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: We could just revert the changes you made to the constructor and the first time a delayed message is called we try to call it and if we get this exception with that specific code, we could just fall back to using the base class behavior for delayed messages. This way it's not an extra option, it just works if configured.

Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: When this exception occurs we could store it in a flag, and then short circuit it if not available so it only happens once.

Comment (barkhashah) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: I introduced a new constructor parameter called "bool delayedExchange" that tells us whether to create delayed exchange or not. Now declaring the exchange happens in the constructor. This variable was the only way that was telling me what kind of exchange to declare ( delayed or actual ) . Now am I understanding this right that you want me to take out "bool delayedExchange" from the constructor and always try to create the delayed exchange and if that fails then create the actual and store the ( exception occured flag for short circuiting future exchange declare as long as its not a new instance of RabbitMQMessageBus) . We come to know only later during publish method that user intends to delay sending the message, however, we create exchange in the constructor itself.

Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Yes, What do people normally do? You'd think that I'd always want a delayed exchange as it gives me both non delayed and delayed messaging support (right?). If we can remove an option and have the best of both worlds (either delayed messages via in process support or offloaded onto rabbitmq) that would be awesome :)

Comment (barkhashah) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: I agree. However, people need to be smart enough to not use the delayed exchange for sending the non-delayed message. There is a peformance issue with that.

Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Looked at that link you post, any disadvantages to having two exchanges defined with the same exact name? If we fail to create or publish to the one we dispose of it and set it to null and call one for no delay, and the other or base implementation for delayed messages?

By doing one or the other right now and taking a perf hit kinda sucks, especially when it might be total random when you want to do a delayed message...

Comment (barkhashah) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Yeah there is a problem if you try to create delayed and then actual exchange or vice versa with the *same name*.

Lets say you start with creating an exchange name as "foundatioExchange" and there is no plugin enabled for delay exchange. Now as per our logic we will always attempt to create x-delayed-type exchange and if that fails we fall back to the actual one. So now we have a successful creation of a regular exchange called "foundatioExchange".

Now you enable the plugin.

After enabling the plugin if you try to create a new RabbitMQMessageBus pub/sub with the same name as "foundatioExchange" you will be thrown with OperationInterruptedException .

+		o	{"The AMQP operation was interrupted: AMQP close-reason, initiated by Peer, code=406, text=\"PRECONDITION_FAILED - inequivalent arg 'type' for exchange 'FoundatioExchange' in vhost '/': received ''x-delayed-message'' but current is 'direct'\", classId=40, methodId=10, cause="}	RabbitMQ.Client.Exceptions.OperationInterruptedException


So you can't keep the same names for 2 different types of exchange. 
I have made some changes where we remove the extra boolean from the constructor and always create the delayed exchange first. If we fail then we fall back to the regular one. Currently testing various schenarios at this point. I will submit for new changes tomorrow.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs (hunk 2 above)
Comment (barkhashah) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Well this is the only unit test case which is failing for me. I am still trying to understand its behavior.

Test Name:	CanSendDelayedMessage
Test FullName:	Foundatio.RabbitMQ.Tests.Messaging.RabbitMqMessageBusTests.CanSendDelayedMessage
Test Source:	C:\Users\barkha.shah\Foundatio\test\Foundatio.RabbitMQ.Tests\Messaging\RabbitMqMessageBusTests.cs : line 39
Test Outcome:	Failed
Test Duration:	0:10:34.256

Result StackTrace:	
at Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer`1 comparer)
   at Foundatio.Tests.Messaging.MessageBusTestBase.<CanSendDelayedMessage>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass46_1.<<InvokeTestMethodAsync>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExecutionTimer.<AggregateAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExceptionAggregator.<RunAsync>d__9.MoveNext()
Result Message:	
Assert.Equal() Failure
Expected: 0
Actual:   998

Comment (barkhashah) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: So when I debug this particular method, I get failure. However, when I run the test seperately it passes.

Comment (barkhashah) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: I have this logic in the PublishAsync now :

`            // if the rabbitmq plugin is not availaible then use the base class delay mechanism
            if (!_delayedExchangePluginEnabled  && delay.HasValue && delay.Value > TimeSpan.Zero) {
                _logger.Trace("Schedule delayed message: {messageType} ({delay}ms)", messageType.FullName, delay.Value.TotalMilliseconds);
                await AddDelayedMessageAsync(messageType, message, delay.Value).AnyContext();
                return;
            }`
When I run the unit test case with the plugin enabled then the test case for CanSendDelayedMessage fails because the above logic will not get executed. With plugin disabled the test case passes.
